Add URI, URL Prisms

n4to4 opened the pull request:
Fix #488 , #489

Maybe there are better/simpler ways to generate `Arbitrary`, `Cogen` instances? Any feedback is welcome :)

xuwei-k commented on test/shared/src/test/scala/monocle/TestInstances.scala:
```diff
@@ -196,4 +199,28 @@ trait TestInstances {
+
+  implicit def urlArbitrary: Arbitrary[URL] = Arbitrary {
+    for {
+      protocol <- Gen.oneOf("http", "https")
```
> Protocol handlers for the following protocols are guaranteed to exist on the search path :- 
>     http, https, file, and jar

https://docs.oracle.com/javase/8/docs/api/java/net/URL.html#URL-java.lang.String-java.lang.String-int-java.lang.String-

n4to4 commented:
@xuwei-k Thanks for your review.

- I moved `java.net.URL` specific code to the `jvm` specific directory.
  (I'm not sure if this is the best way to separate `js`/`jvm` specific code, though :sweat:)
  
- About adding protocols, it seems that `jar` protocol has its own format: `jar:<url>!/{entry}`,
  and `new URL("jar://host.com:123/path/to/jar")` throws an exception, for example.
  So I added a `Gen[URL]` for `jar`.
  https://docs.oracle.com/javase/8/docs/api/java/net/JarURLConnection.html

julien-truffaut commented:
LGTM, what do you think @xuwei-k ?

xuwei-k commented on test/jvm/src/test/scala/monocle/PlatformSpecificTestInstances.scala:
```diff
@@ -0,0 +1,35 @@
+  implicit val urlEqual = Equal.equalA[URL]
+
+  implicit def urlArbitrary: Arbitrary[URL] = Arbitrary {
+    def idGen = Gen.nonEmptyListOf(Gen.alphaChar).map(_.mkString)
```
`val idGen`

xuwei-k commented on test/jvm/src/test/scala/monocle/PlatformSpecificTestInstances.scala:
```diff
@@ -0,0 +1,35 @@
+      path <- path("!/")
+    } yield new URL("jar:" + url + path)
+
+    Gen.frequency((1, url), (1, jar))
```
`Gen.oneOf(url, jar)`

or

`Gen.frequency((3, url), (1, jar))`

?

xuwei-k commented on test/shared/src/test/scala/monocle/TestInstances.scala:
```diff
@@ -196,4 +198,15 @@ trait TestInstances {
     Cogen[(Long, Long)].contramap[UUID]((u: UUID) => (u.getMostSignificantBits, u.getLeastSignificantBits))
 
+  implicit def uriArbitrary: Arbitrary[URI] = Arbitrary {
+    def idGen = Gen.nonEmptyListOf(Gen.alphaChar).map(_.mkString)
```
`val idGen`

xuwei-k commented on test/jvm/src/test/scala/monocle/PlatformSpecificTestInstances.scala:
```diff
@@ -0,0 +1,35 @@
+      path <- Gen.listOfN(length, idGen)
+    } yield path.mkString(sep, "/", "")
+
+    def url: Gen[URL] = for {
```
`val url`

julien-truffaut commented:
Great, thank you guys 👍
